src/FsToolkit.ErrorHandling/Array.fs

```diff
@@ -0,0 +1,244 @@
+
+    let sequenceAsyncResultA xs = traverseAsyncResultA id xs
+
+    let rec private traverseOptionM' (state: Option<_>) (f: _ -> Option<_>) xs =
+        match xs with
+        | [||] ->
+            state
+            |> Option.map Array.rev
+        | arr ->
+            let x = Array.head arr
+            let xs = Array.skip 1 arr
+
+            let r =
+                option {
+                    let! y = f x
+                    let! ys = state
+                    return Array.append [| y |] ys
+                }
+
+            match r with
+            | Some _ -> traverseOptionM' r f xs
+            | None -> r
+
+    let rec private traverseAsyncOptionM' (state: Async<Option<_>>) (f: _ -> Async<Option<_>>) xs =
+        match xs with
+        | [||] ->
+            state
+            |> AsyncOption.map Array.rev
+        | arr ->
+            let x = Array.head arr
+            let xs = Array.skip 1 arr
+
+            async {
+                let! o =
+                    asyncOption {
+                        let! y = f x
+                        let! ys = state
+                        return Array.append [| y |] ys
+                    }
+
+                match o with
+                | Some _ -> return! traverseAsyncOptionM' (Async.singleton o) f xs
+                | None -> return o
+            }
+
+    /// <summary>
+    /// Applies the given function <paramref name="f"/> to each element in the input list <paramref name="xs"/>,
+    /// and returns an option containing a list of the results. If any of the function applications return None,
+    /// the entire result will be None.
+    /// </summary>
+    /// <param name="f">The function to apply to each element in the input list.</param>
+    /// <param name="xs">The input list.</param>
+    /// <returns>An option containing a list of the results of applying the function to each element in the input list,
+    /// or None if any of the function applications return None.</returns>
+    let traverseOptionM f xs = traverseOptionM' (Some [||]) f xs
+
+    /// <summary>
+    /// Applies the monadic function <paramref name="id"/> to each element in the input list <paramref name="xs"/>,
+    /// and returns the result as an option. If any element in the list is None, the entire result will be None.
+    /// </summary>
+    /// <param name="xs">The input list.</param>
+    /// <returns>An option containing the result of applying <paramref name="id"/> to each element in <paramref name="xs"/>.</returns>
+    let sequenceOptionM xs = traverseOptionM id xs
+
+    let traverseAsyncOptionM f xs =
+        traverseAsyncOptionM' (AsyncOption.retn [||]) f xs
+
+    let sequenceAsyncOptionM xs = traverseAsyncOptionM id xs
+
+#if !FABLE_COMPILER
+    let rec private traverseVOptionM' (state: voption<_>) (f: _ -> voption<_>) xs =
+        match xs with
+        | [||] ->
+            state
+            |> ValueOption.map Array.rev
+        | arr ->
+            let x = Array.head arr
```

Comment:
isnt there an array.split or similar that can do this pair of ops one shot?

Reply:
Not in a one shot, it returns 2 arrays, which is not what we want here.

Reply:
true, not sure what I was thinking!

Reply:
All good, no worries :D